docs(madr): add Locality Aware load balancing madr (#8022)
kumahq/kuma: docs/madr/decisions/031-locality-aware-load-balancing.md

# MeshLoadBalancingStrategy - locality aware
* Status: accepted

Technical Story: https://github.com/kumahq/kuma/issues/2689

## Context and Problem Statement
Right now we have really simple locality aware load balancing, we only care about zones. Dataplane will prefer sending traffic to local zone with fallback to other zones. We would like to add more sophisticated mechanism of locality aware load balancing. We would like to allow user to specify load balancing rules based on tags. For example, big cluster is deployed in multiple AZ’s, and we would like to prefer staying in a single AZ to reduce traffic cost.

> Considered Options

Author: fixed

> ## Decision Outcome

Author: this will be filled, when we will have final version

## Solution
As stated earlier, we would like to improve our current locality aware load balancing, and make it more flexible for users.
### Use cases
After gathering client's expectations and a couple of brainstorming sessions, we ended up with multiple use cases for locality aware load balancing:

> 1. Keeping traffic in single zone

Author: yes, with no fallback

2. Load balancing based on priorities within a zone. Example: node → AZ → region
3. Prefer traffic in single zone with fallback to other zones
4. Load balance equally to list of zones
5. When leaving zone, target specific zones first. Example zones physically close to each other
6. Load balance equally to local dpp and other zone
7. Exclude zone from load balancing

### API design
Looking at use cases, it is clear that we have two types of load balancing strategies, inside zone and cross zone. We can easily route to services with proper tags inside the zone, but immediately after we cross zone border and hit Zone Ingress we lose the ability to control locality.
Because of this we have decided that we will split our API in two sections: `inZone` and `crossZone`.

Here is the full example of policy:

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:

>     localityAware:

Author: changed

      inZone: ["k8s.io/node", "k8s.io/az"] # (1)
      crossZone: # (2)
        rules: # (3)
          - groups: # (4)
            - ["zone-1", "zone-2"]

>             - ["zone-3", "zone-4"]

Reviewer: To be future proof does it make sense to nest the array:

```
- groups:
   - zones: ["zone-1", "zone-2"] 
```

That way the day we want to add a field it's not a breaking change

Author: changed

          - groups:
            - ["zone-5"]

>         fallback: any # (5)

Author: changed

```

> (1) In `inZone` section, you configure the priorities of tags used for load balancing.

Reviewer: What happens when list is empty? What happens when it's not set?

Reviewer 2: My expectation is "default behavior" so that every DPP in the zone is load balanced equally just like we would not apply this policy.

Reviewer: I agree

Author: updated mard

(2) In `crossZone` section, you configure zones priority.

(3) In `rules` section, you configure list of zone groups priorities, first groups will have the highest load balancing priority and last in order will have the lowest priority.

> (4) In `groups` section, you configure logical zone groups. Let's assume `zone-1` and `zone-2` are located in the same datacenter, so you want to load balance equally between them. `Zone-5` is located in another datacenter so it should have lower priority, that is why it is in the second groups entry.

Reviewer: Little confused. How do I get routed if I'm a DPP in zone-5?

Reviewer: Also with the first set of groups do I route to any zone regardless of which zone I'm part of or does it still favour self?

Reviewer: What happens when the group is empty?

Reviewer 3: So the inzone traffic has the highest priority. In this example we would have 
Priority 0: the same k8s.io/node
Priority 1: the same k8s.io/az
Priority 2: Zone 1. zone-2
Priority 3: Zone5

If you don't define any crossZone we want to assume that you want to keep traffic within the zone but fallback to any:
```
  defaults:
    localityAwareness:
      inZone: ["k8s.io/node", "k8s.io/az"]
      crossZone: {}
```

When groups are empty we assume you want to treat all zones the same.

```
  defaults:
    localityAwareness:
      inZone: ["k8s.io/node", "k8s.io/az"]
      crossZone:
       rules:
         - groups: []
       
```

Does it make sense?

Author: this example is no longer valid

> (5) At last we should apply some feedback if no dataplanes from previous configuration are available, at this moment we allow `fail` and `any` fallback

Reviewer: ```suggestion
(5) At last we should apply some fallback if no dataplanes from previous configuration are available, at this moment we allow `fail` and `any` fallback
```

Reviewer: What is the value of this when it's not set? any I guess? Also I think technically it should be `Any` or `Fail`

> (5) At last we should apply some fallback if no dataplanes from previous configuration are available, at this moment we allow `fail` and `any` fallback

Reviewer 2: what is the default if not specified?

Author: added default `any`

### API examples based on use cases

> #### Keeping traffic in single zone

Reviewer: ```suggestion
#### Keeping traffic in same zone (disable cross zone traffic)
```

![Use case 1](assets/031/use_case_1.png)

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:

>     localityAware:

Reviewer 3: I think we want to deprecate it and use more advanced rules.

>     localityAware:

Reviewer 3: We are going to keep it for now. We can decide if we want to do something with it later

      inZone: []
```

#### Load balancing based on priorities within a zone. Example: node → AZ → region

![Use case 2](assets/031/use_case_2.png)

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAware:
      inZone: ["k8s.io/node", "k8s.io/az"]
```

> #### Prefer traffic in single zone with fallback to other zones

Reviewer: ```suggestion
#### Prefer traffic in same zone with fallback to any other zone
```

> ![Use case 3](assets/031/use_case_3.png)

Reviewer: The drawings are very helpful though I think you should show all the cases for each of the DPP

Author: I was thinking about this, but I was afraid that it will get messy and harder to actually understand

Reviewer: I think wiht your color codes it should be fine

Author: I've added separate diagram for each ddp to make it clear

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAware:
        inZone: []

>     crossZone:

Author: fixed

      fallback: any
```

#### Load balance equally to list of zones

> ![Use case 4](assets/031/use_case_4.png)

Author: there was typo in example

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAware:
      inZone: []

>     crossZone:

Author: fixed

      - groups:
        - ["zone-1", "zone-3"]
```

#### When leaving zone, target specific zones first

![Use case 5](assets/031/use_case_5.png)

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAware:
      inZone: []
    crossZone:
      - groups:
        - ["zone-1"]
      - groups:
        - ["zone-3"]
```

#### Exclude single zone from load balancing

![Use case 6](assets/031/use_case_6.png)

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAware:
      inZone: []
    crossZone:
      - groups:
        - ["zone-1"]
      fallback: fail
```

> TODO What if i have lots of zones and don’t want to write all of them during the incident. I would like to just exclude single zone, do we want to add something like:

Reviewer 2: You can still set Zone#enabled=false to stop syncing ZoneIngress of this zones to other zones. Therefore excluding the cross zone traffic to this specific zone

Reviewer: But that's for the entire zone and has side effects no? I find being able to exclude 1 zone for 1 service sounds feasible

  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAware:
      inZone: []
    crossZone:
      - exclude: ["zone-3"]
```

#### Load balance equally to local and zone

![Use case 7](assets/031/use_case_7.png)

We have the couple of use cases for this, but no config yet:
- You are migrating from universal to k8s, both zones are in the same AZ and I would like to move traffic gradually from one to another
- You have have some zone with testing or preview environment and you would like to handle live traffic on it for testing. Is it a use case for locality aware lb? Or should this be done with MeshHTTPRoute?
The question is if we really need it, or can this be done currently in some other way?

### Implementation

We want to enhance current `MeshLoadBalancingStrategy` so code related to the changes should be placed in the plugin.

#### Matching endpoints
Each instance of the service might be in a different locality: zone, node. Users need to properly tag these dataplanes so we are able to match them. We need to retrieve all tags from all inbounds and, later, based on these tags, match outbounds that are in the same location.

Example:

> On Kubernetes, all tags are populated automatically, while on Universal, users need to populate tags themselves:

Reviewer: Have we checked this? I'm not sure labels are added to pods I think they are only available on the nodes. https://kubernetes.io/docs/concepts/services-networking/topology-aware-routing/

Reviewer 3: I think I need to rephrase it, it's more about pods tag, thet are in the cluster. I think if a user doesn't have a node tag on the pod we might not get the value.

```yaml
type: Dataplane
mesh: default
name: demo
networking:
  address: 127.0.0.1
  inbound:
    - port: 9999
      servicePort: 10000
      serviceAddress: 127.0.0.1
      tags:
        kuma.io/service: demo
        kuma.io/protocol: tcp
        k8s.io/az: us-east-1
```

> Later, based on these tags, the control-plane creates `ClusterLoadBalancing` and puts all tags, for each endpoint, into cluster metadata::

Reviewer 4: Do you mean `ClusterLoadAssignment`?

```json
"metadata":{
  "filter_metadata": {
    "envoy.lb": {
      "az/test": "b",
      "kuma.io/protocol": "tcp",
      "k8s.io/az": "us-east-1"
    }
  }
}
```

We can retrieve them from `ClusterLoadBalancing` and match them with the policy. First, we need to retrieve all tags from inbounds so we can determine our location and match it with endpoints.

Pseudo algorithm:

1. Get all inbound tags.
2. Take the matched rules.
3. Check if inbound has the tag defined in the rule. 
4. If yes, iterate over all endpoints of the service and group them in to priority groups, by matching with the specific rule( first inZone, then  cross Zone). If no, get all the endpoints in the zone. 
5. Create `LocalityLbEndpoints` with selected endpoints. 
6. Override `ClusterLoadAssignment` for the dataplane in the `ResourceStore`. 

Dataplanes in the specific location are going to receive only endpoints in the specific locality. For cross zone traffic without egress the control-plane delivers only ingresses of specific zones.

##### Egress

> Egress is not as simple as ingress. Currently, we support Locality Aware when atleast one client requires it, so we cannot distinguish between clients. Control-plane needs to send all dataplanes to the egress sidecar because there might be services sending requests to all zones. There are 2 options:

Reviewer 4: nit: 
```suggestion
Egress is not as simple as ingress. Currently, we support Locality Aware when atleast one client requires it, so we cannot distinguish between clients. Control-plane needs to send all dataplanes to the egress because there might be services sending requests to all zones. There are 2 options:
```

- LoadBalacing based on metadata
- Separate clusters

Load balancing based on the metadata seems to be able to solve an issue but it has limitations. Because, egress knows all endpoints and does routing based on matching metadata the request might not obey the priority of zones. In this case we cannot configure different priorities for different clients.

Separate cluster seems to be the only option that could solve this problem. When there is a different configuration of Load Balancing for the one of dataplane, control-plane sends to egress new a cluster `mesh-name:service-name:hash(tag + value)` with the routing based on the SNI: `service-name{mesh=mesh-name,hash=hash}`

> This cluster has a different configuration of priorities and only subset of zones. Dataplanes interested in this Load Balancing are going to have the same configuration of the cluster which sets SNI: `service-name{mesh=mesh-name,hash=hash}` and based in this SNI egress can make a routing decision to route to the specific cluster.

Reviewer: Could we just say that Egress will be the equivalent of top level targetRef be mesh and won't make a difference between client?

It's likely an acceptable shortcoming. I think down the line @jakubdyszkiewicz mentioned maybe having one cluster per zone (and route) which would make these things a bit easier maybe.

Reviewer 3: Yes, cluster/route per zone would be simpler in this case because we could deliver just these clusters.

##### Ingress

> Ingress should receive all endpoints, because there is no rules for in zone traffic when the request comes from the outside.

Reviewer 2: I cannot figure this out after reading the doc. Do we apply any load balancing rules on the Ingress when Ingress is proxying to the DPPs?

##### Serviceless
On kubernetes you can create a deployment without the service, in this case control-planes creates default inbound so tags should be available.

> On universal you cannot create a dataplane without inbound or gateway definition.

Reviewer: Do we even care? You can't LoabBalance to a service without an inbound right?

Reviewer 3: Yes, you cannot route to it but you need to route from it. To figure out what is your location the easiest way is to use Inbound tags.

Reviewer: Ha ok. I think the sentence is a little confusing here. I think there's really no need to mention this, it's just what we do in matching of top level targetRefs there's nothing different in this MADR

#### ClusterLoadBalancing cache

> Currently, while retrieving endpoints in the plugin, we are getting a pointer to the object in the cache that is shared between dataplanes. This won't work with LocalityAware load balancing because each dataplane might have a different subset of endpoints with different priorities. In the first iteration, we would like to try using it without CLA cache and measure the impact of the policy in the bigger cluster with many nodes. If the impact of not using the cache is significant, we might have to keep this configuration in the cache if there are more services on the same node. The cache might not improve the performance if there are instances of the same service in different localities (nodes, availability zones).

Reviewer: @lobkovilya and @jakubdyszkiewicz can probably already tell you that we can't live without the CLA cache and that's why it was introduced.

However the CLA cache key is:

`key := sha256.Hash(fmt.Sprintf("%s:%s:%s:%s", apiVersion, meshName, cluster.Hash(), meshHash))`

You could quite easily keep a CLA per `rule` and still have the cache.

I think picking the right rule is probably pretty quick (especially as in most cases there will only be a top level targetRef of `kind:Mesh`)

Reviewer 2: It's hard to say if we really can't live without CLA cache. That was the first optimization for CP perf even before MeshCache and way way before reachable services. If you use reachable services, I'm not sure if the CLA cache is that critical. We need to measure it.

Reviewer 2: Maybe I'm reading it wrong, but I think your cache key wouldn't work, because you need to also add "input for your load balancing policy" to the key. For example. You may have same policy to prefer `k8s.io/node`, but DPP on different nodes would have CLA with different priorities.

Author: we will deal with CLA cache later on as a separate issue

### Other
#### Cross mesh
We cannot configure this policy cross mesh. We are bound to mesh, when you leave mesh you lose all information needed for locality aware load balancing, and you need to specify new policy for each mesh.

> Question what if someone is placing each bounded context in a different mesh for isolation and would still like to benefit from locality aware load balancing? Will cross mesh traffic be routed locally? 

Reviewer: WIP :) 

Ideally yes you'd like to pick a cross mesh gateway that's close. Not sure how this works though...

Reviewer 2: Cross mesh gateway, in the end, is "yet another destination with tags". It hides underlying DPPs tags (for a good reason!), but we can add standard k8s node tags, zone tags, and whatever tags a user will add to it.

5. Create `LocalityLbEndpoints` with selected endpoints. 
6. Override `ClusterLoadAssignment` for the dataplane in the `ResourceStore`. 

Dataplanes in the specific location are going to receive only endpoints in the specific locality. For cross zone traffic without egress the control-plane delivers only ingresses of specific zones.

##### Egress
Egress is not as simple as ingress. Currently, we support Locality Aware when atleast one client requires it, so we cannot distinguish between clients. Control-plane needs to send all dataplanes to the egress sidecar because there might be services sending requests to all zones. There are 2 options:
- LoadBalacing based on metadata
- Separate clusters

Load balancing based on the metadata seems to be able to solve an issue but it has limitations. Because, egress knows all endpoints and does routing based on matching metadata the request might not obey the priority of zones. In this case we cannot configure different priorities for different clients.

> Separate cluster seems to be the only option that could solve this problem. When there is a different configuration of Load Balancing for the one of dataplane, control-plane sends to egress new a cluster `mesh-name:service-name:hash(tag + value)` with the routing based on the SNI: `service-name{mesh=mesh-name,hash=hash}`

Reviewer 4: I'm not sure I understand how we want to generate separate clusters. Is it a cluster per client or per destination?

Reviewer 3: Cluster per client. Each client might define different priorities for cross zone communication. 
Servica A [zone-1 : 1, zone-2: 2]
Service B [zone2: 1]
In the case of egress, it knows about all zones so if we are going to use one cluster we can't ensure that traffic is routed based on the provided definition for the client.

Reviewer 4: What about SNI? IIRC ZoneEgress doesn't rewrite SNI at this moment and ZoneIngress later is using the same SNI in filterChainMatches. 
1. Is it possible to rewrite SNI without TLS termination?
2. If it's not possible, does it mean ZoneIngress has to generate filterChainMatches using clients' hashes?

Reviewer 3: 1. No, it isn't
2. Yes, ingresses would need filter chains with hashes to make it work.

Reviewer: > Is it possible to rewrite SNI without TLS termination?
Why do we need to rewrite SNI?

Reviewer 3: 1. Egress has cluster definition `service-a`, when `client-1` wants to make a request to `service-a` but only in `zone-1` and `zone-2` but client `client-2` wants to make a request to `service -a` in `zone-3` and `zone-4` with lower priority egress won't use client logis but global one.
2. Because when we override SNI ingress needs to match SNI from the request and in this case it might not be the same

Reviewer 4: I'm worried about the scalability of this approach. Imagine a bit extreme but still possible use case when a user creates MLBS for each service:
```yaml
targetRef:
  kind: MeshService
  name: "${service}"
to:
  targetRef:
    kind: Mesh
  defaults:
    localityAwareness: {...}
```

In that case, the number of filterChainMatches in ZoneIngress is `num_services*num_of_destinations`, while today it's `num_of_destinations`

Reviewer 2: Some thoughts that I shared with @lukidzi thinking out loud on priv.

Keep in mind that AFAIU for crosszone routing we only care about routing by zone.

I also don't think separate clusters on egress are the way to go.
Let's say we have a service in zone1 (my local), zone2 zone3, zone4 and we only want to fallback to zone2+zone3.

It's easy with Ingress, because we set priorities
local endpoint - zone1: 0
zone ingress of zone2: 1
zone ingress of zone3: 1

It's hard with Egress, because all zone ingresses are hidden behind one zone egress.
If we solve the problem that we can put multiple endpoints with different metadata we could do this:
local endpoint - zone1: 0
zone egress: 1 that sends SNI: `serviceA{mesh=default,kuma.io/zone=zone2}`
zone egress: 1 that sends SNI: `serviceA{mesh=default,kuma.io/zone=zone3}`

At this moment we cannot do this because Envoy limitation, but it's not the first time it's a problem for us. Maybe we could solve this.

Alternatively, we could consider separate clusters but on the client side and try to use https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/aggregate_cluster

Reviewer 3: I've included @jakubdyszkiewicz's suggestion with all the details. I feel like it's possible but need to confirm the aggregate clusters part.

# MeshLoadBalancingStrategy - locality aware
* Status: accepted

Technical Story: https://github.com/kumahq/kuma/issues/2689

## Context and Problem Statement
Right now we have really simple locality aware load balancing, we only care about zones. Dataplane will prefer sending traffic to local zone with fallback to other zones. We would like to add more sophisticated mechanism of locality aware load balancing. We would like to allow user to specify load balancing rules based on tags. For example, big cluster is deployed in multiple AZ’s, and we would like to prefer staying in a single AZ to reduce traffic cost.

## Considered Options

## Decision Outcome

## Solution
As stated earlier, we would like to improve our current locality aware load balancing, and make it more flexible for users.
### Use cases
After gathering client's expectations and a couple of brainstorming sessions, we ended up with multiple use cases for locality aware load balancing:

> 1. Keeping traffic in single zone

Reviewer: ```suggestion
1. Keeping traffic in single zone (disable all cross-zone traffic)
```

As Ilya asked for clarification let's make this explicit

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: ["k8s.io/node", "k8s.io/az"] # (1)
      crossZone: # (2)
        rules: # (3)
          - groups: # (4)
            - zones: ["zone-1", "zone-2"]
            - zones: ["zone-3", "zone-4"]
          - groups:
            - zones: ["zone-5"]
        fallback: Any # (5)
```

(1) In `inZone` section, you configure the priorities of tags used for load balancing. If list is empty or inZone is not specified, load balancing works normally, without taking any priorities into account. 

(2) In `crossZone` section, you configure zones priority.

(3) In `rules` section, you configure list of zone groups priorities, first groups will have the highest load balancing priority and last in order will have the lowest priority.

(4) In `groups` section, you configure logical zone groups. Let's assume `zone-1` and `zone-2` are located in the same datacenter, so you want to load balance equally between them. `Zone-5` is located in another datacenter so it should have lower priority, that is why it is in the second groups entry.

> (5) At last we should apply some fallback if no dataplanes from previous configuration are available, at this moment we allow `Fail` and `Any` fallback. Default value `Any`'

Reviewer: Should default be `Fail` to make enabling mesh on a k8s cluster no have any surprising behavior?

Reviewer 3: Make sense, when the user deploys the mesh and connects the second zone, without any configuration traffic, currently traffic is routed to all services equally. That would make it more k8s way.

### API examples based on use cases

#### Keeping traffic in same zone (disable cross zone traffic)

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: []
```

![Use case 1](assets/031/use_case_1.png)

#### Load balancing based on priorities within a zone. Example: node → AZ → region

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: ["k8s.io/node", "k8s.io/az"]
```

> ![Use case 2](assets/031/use_case_2.png)

Reviewer: ```suggestion
![Use case 2](assets/031/use_case_2_1.png)
```

> ![Use case 2](assets/031/use_case_2.png)

Reviewer: ```suggestion
![Use case 2](assets/031/use_case_2_1.png)
```

Reviewer: Same for the 2 other entries

![Use case 2](assets/031/use_case_2.png)
![Use case 2](assets/031/use_case_2.png)

#### Prefer traffic in same zone with fallback to any other zone

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: []
      crossZone:
        fallback: Any
```

![Use case 3](assets/031/use_case_3_1.png)
![Use case 3](assets/031/use_case_3_2.png)

> #### Load balance equally to list of zones

Reviewer: Worth adding a text explanation to say that zone-2 is excluded so traffic will never come from the outside to zone 2.

It's clear from the images but I think text would help

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: []
      crossZone:
        - groups:
          - zones: ["zone-1", "zone-3"]
        fallback: Fail
```

![Use case 4](assets/031/use_case_4_1.png)
![Use case 4](assets/031/use_case_4_2.png)
![Use case 4](assets/031/use_case_4_3.png)

#### When leaving zone, target specific zones first

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: []
      crossZone:
        - groups:
          - zones: ["zone-1"]
        - groups:
          - zones: ["zone-3"]
```

![Use case 5](assets/031/use_case_5_1.png)
![Use case 5](assets/031/use_case_5_2.png)
![Use case 5](assets/031/use_case_5_3.png)

> #### Exclude single zone from load balancing

Reviewer: The description is not really true. If I add `zone-4` it will also be excluded. I think we don't really have a way to exclude just a set of zone atm.

Reviewer: Actually just saw there's further comments lower down

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: []
      crossZone:
        - groups:
          - zones: ["zone-1"]
        fallback: Fail
```

![Use case 6](assets/031/use_case_6_1.png)
![Use case 6](assets/031/use_case_6_2.png)
![Use case 6](assets/031/use_case_6_3.png)

TODO What if i have lots of zones and don’t want to write all of them during the incident. I would like to just exclude single zone, do we want to add something like:

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: []
      crossZone:
        - exclude:

>             zones: ["zone-3"]

Reviewer: Does it make sense to exclude zone at a level? Isn't the exclude just like `fallback` ? 

So we'd have:

```
crossZone:
  groups:
     - zones: ["zone-1", "zone-2"]
   fallback:
      type: Any
      exclude:
         zones: ["zone-4"]
```

Reviewer 3: I've changed the model. 

```
crossZone:
  - groups:
     - name: eu
       zones: ["zone-1", "zone-2"]
     - name: us
       zones: ["zone-3"]
  - groups:
    - name: eu
      zones: ["zone-5"]
   fallback:
      type: Any
      exclude:
         zones: ["zone-4"]
```

this model allows to match groups in case of more than one priority for each group

```

#### Load balance equally to local and zone

![Use case 7](assets/031/use_case_7.png)

We have the couple of use cases for this, but no config yet:

> - You are migrating from universal to k8s, both zones are in the same AZ and I would like to move traffic gradually from one to another

Reviewer 3: I think we can expose it. It makes sense to define and allow users to configure it.

- You have have some zone with testing or preview environment and you would like to handle live traffic on it for testing. Is it a use case for locality aware lb? Or should this be done with MeshHTTPRoute?
The question is if we really need it, or can this be done currently in some other way?

> 

Reviewer: We're missing an example. The case where I have multiple sets of zones that I don't want to overlap.

For example I have us-east and us-west and the other side is eu-west and eu-east.

What I want is "us-east/west" failover and "eu-east/west" failover with no risk of having cross ocean traffic.

Originnaly we thought about writing this:

```yaml
crossZones:
  groups:
     - zones: ["us-east", "us-west"]
     - zones: ["eu-east", "eu-west"]
  fallback: Fail
```

but I feel it goes against other examples here.

### Implementation

We want to enhance current `MeshLoadBalancingStrategy` so code related to the changes should be placed in the plugin.

#### Matching endpoints
Each instance of the service might be in a different locality: zone, node. Users need to properly tag these dataplanes so we are able to match them. We need to retrieve all tags from all inbounds and, later, based on these tags, match outbounds that are in the same location.

Example:

> On Kubernetes, all tags from the pod are populated automatically, while on Universal, users need to populate tags themselves:

Reviewer: We talked about it with @lukidzi we'd need a controller on k8s that reconciles the DPP to bubble the Node tags to the Dataplane object.

I think this should be explained here and indicate that writing this controller is out of scope of the first release of this feature

Reviewer 3: True, I've added it but a bit lower: https://github.com/kumahq/kuma/pull/8022/files#diff-ce8ace61d35c2efd60262c423f5763d63fb047d8d7b000deefa90beb647bb36aR269 I will move it higher

Reviewer 3: Moved to separate section

```yaml
type: Dataplane
mesh: default
name: demo
networking:
  address: 127.0.0.1
  inbound:
    - port: 9999
      servicePort: 10000
      serviceAddress: 127.0.0.1
      tags:
        kuma.io/service: demo
        kuma.io/protocol: tcp
        k8s.io/az: us-east-1
```

Later, based on these tags, the control-plane creates `ClusterLoadAssignment` and puts all tags, for each endpoint, into cluster metadata::

```json
"metadata":{
  "filter_metadata": {
    "envoy.lb": {
      "az/test": "b",
      "kuma.io/protocol": "tcp",
      "k8s.io/az": "us-east-1"
    }
  }
}
```

We can retrieve them from `ClusterLoadAssignment` and match them with the policy. First, we need to retrieve all tags from inbounds so we can determine our location and match it with endpoints.

Pseudo algorithm:

1. Get all inbound tags.
2. Take the matched rules.
3. Check if inbound has the tag defined in the rule. 

> 4. If yes, iterate over all endpoints of the service and group them in to priority groups, by matching with the specific rule( first inZone, then  cross Zone). If no, get all the endpoints in the zone. 

Reviewer: So by default cross Zone traffic would be off? I think it's an ok behaviour especially as this is compliant with regular k8s behaviour

Reviewer: I think we really need to cover details around provision factor etc here. Especially if there are endpoints we could technically route to that we don't send down to Envoy (I'm not sure this can happen though :) ).

Reviewer 3: Do you mean Envoy's provision factor or maybe how we are going to choose endpoints to ensure there are endpoints?

5. Create `LocalityLbEndpoints` with selected endpoints. 
6. Override `ClusterLoadAssignment` for the dataplane in the `ResourceSet`.

Dataplanes in the specific location are going to receive only endpoints in the specific locality. For cross zone traffic without egress the control-plane delivers only ingresses of specific zones.

Question:
* Should we populate the 'node' and 'az' tags from Kubernetes resources to the dataplane object, or should users be responsible for providing them explicitly? 

>   * Some of these informations are available in the 'node' object, and we could retrieve node-related information from Kubernetes and use it to populate tags for the pod. To achieve this, we could implement a Kubernetes controller. To get node info there should be no problem because it's included in Pod object, it's more difficult to retrieve node labels/annotation without doing K8s call for Node object.

Reviewer 3: Rephrased

This cluster has a different configuration of priorities and only subset of zones. Dataplanes interested in this Load Balancing are going to have the same configuration of the cluster which sets SNI: `service-name{mesh=mesh-name,hash=hash}` and based in this SNI egress can make a routing decision to route to the specific cluster. That also requries each ingress to create a filter chain match for SNIs, to match an incoming request with the cluster.

The last option doesn't create separate configuration for each client, but uses just Mesh scoped `MeshLoadBalancingStrategy` to configure egress. That reduces possibilites of configuring Egress but can be changed in the future.

##### Ingress
We don't configure ingresses.

##### Serviceless
On kubernetes you can create a deployment without the service, in this case control-planes creates default inbound so tags should be available.
On universal you cannot create a dataplane without inbound or gateway definition.

#### ClusterLoadBalancing cache

> Currently, while retrieving endpoints in the plugin, we are getting a pointer to the object in the cache that is shared between dataplanes. This won't work with localityAwareness load balancing because each dataplane might have a different subset of endpoints with different priorities. In the first iteration, we would like to try using it without CLA cache and measure the impact of the policy in the bigger cluster with many nodes. If the impact of not using the cache is significant, we might have to keep this configuration in the cache if there are more services on the same node. The cache might not improve the performance if there are instances of the same service in different localities (nodes, availability zones).

Reviewer: How about saying:

- We're going to run tests with Mesh-Perf to see if the CLA cache has a real impact
- If results are good we'll remove the CLA cache
- Otherwise it's likely that the CLA cache will need to be rewritten to have different clusters depending on the dataplane tags.

> 

Reviewer: I think one thing we're missing is explaining what happens when other policies create new clusters. I'm thinking for example of MeshHTTPRoute.

Reviewer 3: I've added note about MeshHttpRoute that would apply for others

6. Load balance equally to local dpp and other zone
7. Exclude zone from load balancing
8. Load balance equally within a group of zones

### API design
Looking at use cases, it is clear that we have two types of load balancing strategies, inside zone and cross zone. We can easily route to services with proper tags inside the zone, but immediately after we cross zone border and hit Zone Ingress we lose the ability to control locality.
Because of this we have decided that we will split our API in two sections: `inZone` and `crossZone`.

Here is the full example of policy:

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: ["k8s.io/node", "k8s.io/az"] # (1)
      crossZone: # (2)
        rules: # (3)
          - groups: # (4)
            - name: "group1" # (5)
              zones: ["zone-1", "zone-2"]
            - name: "group2"
              zones: ["zone-3", "zone-4"]
          - groups:
            - zones: ["zone-5"]
        fallback: # (6)
          type: Any # (7)
          exclude: # (8)
            zones: ["zone-6"]

>       overprovisioningFactor: 120% # (9)

Reviewer: this as `%` doesn't look good. We should just have it as either: `overprovisionPercentage` or factor and have a float

Reviewer 2: We have a precentage in other places in the API (for example MeshFaultInjections I think?), let's be consistent how we expose this to the users

Author: changed to version consistent with MeshFaultInjection

```

> (1) In `inZone` section, you configure the priorities of tags used for load balancing. If list is empty or inZone is not specified, load balancing works normally, without taking any priorities into account. 

Reviewer 2: this is a bit nit, but what do you think about `localZone`. It would be good to have terminology to name the zone in which you route the traffic. Also it would be good to explicitly say what is the value - it is a "tag" which is a base concept in Kuma.
```
localZone:
- tag: "k8s.io/node"
- tag: "k8s.io/az"
```

Reviewer: localZone is for sure much better than `inZone` :)

(2) In `crossZone` section, you configure zones priority.

(3) In `rules` section, you configure list of zone groups priorities, first groups will have the highest load balancing priority and last in order will have the lowest priority.

(4) In `groups` section, you configure logical zone groups. Let's assume `zone-1` and `zone-2` are located in the same datacenter, so you want to load balance equally between them. `Zone-5` is located in another datacenter so it should have lower priority, that is why it is in the second groups entry.

> (5) In the `name` section, you can assign a name to the group, which is later used for matching when there are multiple groups. If no group name is provided, the values from the group will apply to all entries in previous groups.

Reviewer 5: Where is this explained? AFAICT the name is unnecessary

> (5) In the `name` section, you can assign a name to the group, which is later used for matching when there are multiple groups. If no group name is provided, the values from the group will apply to all entries in previous groups.

Reviewer 2: Forgive me if I'm missing something, this MADR is quite complex, but I can't find the example where this group name is used.

  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: []
```

![Use case 1](assets/031/use_case_1.png)

#### Load balancing based on priorities within a zone. Example: node → AZ → region

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: ["k8s.io/node", "k8s.io/az"]
```

![Use case 2](assets/031/use_case_2_1.png)
![Use case 2](assets/031/use_case_2_2.png)
![Use case 2](assets/031/use_case_2_3.png)

#### Prefer traffic in same zone with fallback to any other zone

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: []
      crossZone:
        fallback:
          type: Any
```

![Use case 3](assets/031/use_case_3_1.png)
![Use case 3](assets/031/use_case_3_2.png)

#### Load balance equally to list of zones
In this case, traffic from outside to the `backend` service in `zone-2` is completely excluded.

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: []

>       crossZone:

Reviewer 5: some of the example are missing the `rules` key

        - groups:
          - name: "group1"
            zones: ["zone-1", "zone-3"]
        fallback:
          type: Fail
```

![Use case 4](assets/031/use_case_4_1.png)
![Use case 4](assets/031/use_case_4_2.png)
![Use case 4](assets/031/use_case_4_3.png)

#### When leaving zone, target specific zones first

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: []
      crossZone:
        - groups:
          - zones: ["zone-1"]
            name: "group1"
        - groups:
          - zones: ["zone-3"]
            name: "group1"

> ```

Reviewer 2: Would I achieve the same thing with this YAML
```
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: []
      crossZone:
        - groups:
          - zones: ["zone-1"]
            name: "group1"
          - zones: ["zone-3"]
            name: "group2"
```
?

Reviewer 3: No that would be different, now thinking about this I feel like `groups` are not a great idea in this shape. How would you match that only services in zone `zone-5` should route to `zone-1` but services in `zone-5` should route `zone-3`. I think we are missing a matching mechanism that would say that, or we could use matching by `MeshServiceSubset` and instead of many entries in the group at the same priority level, we could do only one List for a single priority.

![Use case 5](assets/031/use_case_5_1.png)
![Use case 5](assets/031/use_case_5_2.png)
![Use case 5](assets/031/use_case_5_3.png)

#### Exclude single zone from load balancing

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: []
      crossZone:
        fallback:
          type: Any
          exclude:
            zones: ["zone-3"]
```

![Use case 6](assets/031/use_case_6_1.png)
![Use case 6](assets/031/use_case_6_2.png)
![Use case 6](assets/031/use_case_6_3.png)

#### Load balance equally within a group of zones

When we want to keep traffic in specific groups and don't route fallback from EU to US.

![Use case 8](assets/031/use_case_8.png)

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: []

>       crossZone:

Reviewer 2: nit: you're missing rules here and in other examples

        - groups:
          - name: "us"
            zones: ["us-east", "us-west"]
          - name: "eu"
            zones: ["eu-east"]
        - groups:
          - name: "eu"
            zones: ["eu-east2"]
        fallback:
          type: Fail

> ```

Reviewer 2: I'm struggling to understand this example. Why Zone US does not fallback to eu-east2, but Zone EU does?

What is the relation between items in rules and items in groups?

Reviewer 3: We can have many groups with different priorities, I assume that the user can define a mesh policy for all services differently than for others.

So we have a group with Priority 1, one in us and the other in eu. 
But the user has also another zone in eu and might want to fallback to `eu-east2` with priority 2 but fail for us.

## Considered Options

## Decision Outcome

## Solution
As stated earlier, we would like to improve our current locality aware load balancing, and make it more flexible for users.
### Use cases
After gathering client's expectations and a couple of brainstorming sessions, we ended up with multiple use cases for locality aware load balancing:

1. Keeping traffic in single zone (disable all cross-zone traffic)
2. Load balancing based on priorities within a zone. Example: node → AZ → region
3. Prefer traffic in single zone with fallback to other zones
4. Load balance equally to list of zones
5. When leaving zone, target specific zones first. Example zones physically close to each other
6. Load balance equally to local dpp and other zone
7. Exclude zone from load balancing
8. Load balance equally within a group of zones

### API design
Looking at use cases, it is clear that we have two types of load balancing strategies, inside zone and cross zone. We can easily route to services with proper tags inside the zone, but immediately after we cross zone border and hit Zone Ingress we lose the ability to control locality.

> Because of this we have decided that we will split our API in two sections: `inZone` and `crossZone`.

Reviewer 4: ```suggestion
Because of this we have decided that we will split our API in two sections: `localZone ` and `crossZone`.
```

Here is the full example of policy:

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:

>     localityAwareness:

Reviewer 4: I left a comment but it constantly shows as "hidden", just wanted to bump it up https://github.com/kumahq/kuma/pull/8022#discussion_r1356588009

      localZone:
        affinityTags: ["k8s.io/node", "k8s.io/az"]          # (1)
      crossZone:                                            # (2)
        failover:                                           # (3)
          - from:                                           # (4)
              zones: ["zone-1", "zone-2"]
            to:                                             # (5)
              type: Only                                    # (6)
              zones: ["zone-1", "zone-2"]
          - from:
              zones: ["zone-3"]
            to:

>               type: AnyExcept

Reviewer 3: It's not as clear. By default a type is `None` so that makes it clear

>               type: AnyExcept

Reviewer 5: but IMO all are good enough

```

![Use case 1](assets/031/use_case_1.png)

#### Load balancing based on priorities within a zone. Example: node → AZ → region

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      localZone:
        affinityTags: ["k8s.io/node", "k8s.io/az"]
```

![Use case 2](assets/031/use_case_2_1.png)
![Use case 2](assets/031/use_case_2_2.png)
![Use case 2](assets/031/use_case_2_3.png)

#### Prefer traffic in same zone with fallback to any other zone

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      localZone:
        affinityTags: []
      crossZone:
        failover:
          - to: 
              type: Any
```

![Use case 3](assets/031/use_case_3_1.png)
![Use case 3](assets/031/use_case_3_2.png)

#### Load balance equally to list of zones
In this case, traffic from outside to the `backend` service in `zone-2` is completely excluded.

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      localZone:
        affinityTags: []
      crossZone:

>         failover:

Reviewer 4: The word "failover" implies the behavior in this section applies to situations when something went wrong (i.e. requests to local instance started failing and we're looking for an alternative). But the use case "Load balance equally to list of zones" means sending requests between local/zone-1/zone-3 even if everything works as expected

Reviewer 3: I think the description is misleading. We always route to local first and if that fails we are going out of the local zone.

Maybe it should be:
```Load balance equally to list of zones when local not available````

Reviewer 3: I think that the case, maybe in this case you should use Mesh*Route https://github.com/kumahq/kuma/pull/8022/files#diff-ce8ace61d35c2efd60262c423f5763d63fb047d8d7b000deefa90beb647bb36aR250

Reviewer: Yeah it's also odd you would prefer more complex request routing if you can fullfill it locally.

Reviewer 4: Also, what policy takes precedence? If I have MeshLoadBalancingStrategy that prioritizes Local, but then I created MeshHTTPRoute like:

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshHTTPRoute
metadata:
  name: route-all-zones
  namespace: kuma-system
  labels:
    kuma.io/mesh: default
spec:
  targetRef:
    kind: MeshService
    name: frontend
  to:
  - targetRef:
      kind: MeshService
      name: backend
    rules:
      default:
        backendRef:
        - kind: MeshServiceSubset
          name: backend
          tags:
            kuma.io/zone: zone-1
        - kind: MeshServiceSubset
          name: backend
          tags:
            kuma.io/zone: zone-3
```

What will be the behavior?

Reviewer: @lobkovilya does this address your concerns?

Reviewer 3: Order of plugins is `Mesh*Route` and after it `MeshLoadBalancingStrategy`
So when you create
```
apiVersion: kuma.io/v1alpha1
kind: MeshHTTPRoute
metadata:
  name: route-all-zones
  namespace: kuma-system
  labels:
    kuma.io/mesh: default
spec:
  targetRef:
    kind: MeshService
    name: frontend
  to:
  - targetRef:
      kind: MeshService
      name: backend
    rules:
      default:
        backendRef:
        - kind: MeshServiceSubset
          name: backend
          tags:
            kuma.io/zone: zone-1
        - kind: MeshServiceSubset
          name: backend
          tags:
            kuma.io/zone: zone-3
```

later `MeshLoadBalancingStrategy` takes endpoints and configures them based on the routing policy. So if your MLBS allows to route to these zones you will route there

Reviewer 4: what MLBS I have to create to make `route-all-zones` MHR work as expected?

Reviewer 4: I see, having MLBS like this
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshLoadBalancingStrategy
metadata:
  name: mlbs-any
  namespace: kuma-system
  labels:
    kuma.io/mesh: default
spec:
  targetRef:
    kind: Mesh
  to:
    targetRef:
      kind: MeshService
      name: backend
    defaults:
      localityAwareness:
        crossZone:
          failover:
            - to:
                type: Any
```
allows me to create any possible split in MHR.

1. Keeping traffic in single zone (disable all cross-zone traffic)
2. Load balancing based on priorities within a zone. Example: node → AZ → region
3. Prefer traffic in single zone with fallback to other zones
4. Load balance equally to list of zones
5. When leaving zone, target specific zones first. Example zones physically close to each other
6. Load balance equally to local dpp and other zone
7. Exclude zone from load balancing
8. Load balance equally within a group of zones

### API design
Looking at use cases, it is clear that we have two types of load balancing strategies, inside zone and cross zone. We can easily route to services with proper tags inside the zone, but immediately after we cross zone border and hit Zone Ingress we lose the ability to control locality.
Because of this we have decided that we will split our API in two sections: `localZone ` and `crossZone`.

Here is the full example of policy:

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:

>       disabled: false                                       # DEPRECATED

Reviewer: Should we just keep it if someone wants to disable something set of the parent?

        affinityTags: ["k8s.io/node", "k8s.io/az"]          # (1)
      crossZone:                                            # (2)
        failover:                                           # (3)
          - from:                                           # (4)
              zones: ["zone-1", "zone-2"]
            to:                                             # (5)
              type: Only                                    # (6)
              zones: ["zone-1", "zone-2"]
          - from:
              zones: ["zone-3"]
            to:
              type: AnyExcept
              zones: ["zone-1"]
          - to:
              type: Any
      overprovisioningFactor: 
        percentage: 120                                     # (7)
```

(1) In `localZone` section, you configure the affinity of tags used for load balancing. If list is empty or localZone is not specified, load balancing works normally, without taking any priorities into account. 

(2) In `crossZone` section, you configure zones priority.

(3) In `failover` section, you configure list of zone priorities, first rules will have the highest load balancing priority and last in order will have the lowest priority.

> (4) In `from` section, you configure to which zones this rule applies. This rule will apply to all DPPs in `zone-1` and `zone-2`

Reviewer: If empty it means all zones or do we make a semantic difference between `empty` and `omitted` ?

Reviewer 3: I've added the default. I think empty from doesn't make sense so we should make `from.zones: []` empty list. If empty object then all zones

(5) In `to` section, you configure to which zones traffic should be redirected.

(6) `type` field lets you control how fallback zones are picked. Allowed values: 
- `Only` - will load balance traffic only to zones listed in `zones` field 
- `None` - will not load balance traffic any further
- `Any` - will load balance traffic to all zones

> - `AnyExcept` - will load balance traffic to all zones except ones specified in `zones` field

Reviewer: What's the default type? I think it's `None` but we should make it explicit

Reviewer 3: Yes, `None` is the default I will add the info

> (9) In [`overprovisioningFactor`](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/overprovisioning#arch-overview-load-balancing-overprovisioning-factor) section, you configure how early `Envoy` should consider other priorities.

Reviewer 3: The default is 140% in Envoy and I think we can set the same value. I am going to add it

### API examples based on use cases

#### Keeping traffic in same zone (disable cross zone traffic)

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      localZone: 
        affinityTags: []

> ```

Reviewer: Worth making it explicit this is the same as:

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults: {}
```

![Use case 4](assets/031/use_case_4_3.png)

#### When leaving zone, target specific zones first

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      localZone:
        affinityTags: []
      crossZone:
        failover:
          - to:
              type: Only
              zones: ["zone-1"]
          - to:
              type: Only
              zones: ["zone-3"]
```

![Use case 5](assets/031/use_case_5_1.png)
![Use case 5](assets/031/use_case_5_2.png)
![Use case 5](assets/031/use_case_5_3.png)

#### Exclude single zone from load balancing

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      localZone:
        affinityTags: []
      crossZone:
        failover:
          - to:
              type: AnyExcept
              zones: ["zone-3"]
```

![Use case 6](assets/031/use_case_6_1.png)
![Use case 6](assets/031/use_case_6_2.png)
![Use case 6](assets/031/use_case_6_3.png)

#### Load balance equally within a group of zones

When we want to keep traffic in specific groups and don't route fallback from EU to US.

> ![Use case 8](assets/031/use_case_8.png)

Reviewer: If it doesn't make a diff I'd keep the simplest option :)

Example:

On Kubernetes, all tags from the pod are populated automatically, while on Universal, users need to populate tags themselves:

```yaml
type: Dataplane
mesh: default
name: demo
networking:
  address: 127.0.0.1
  inbound:
    - port: 9999
      servicePort: 10000
      serviceAddress: 127.0.0.1
      tags:
        kuma.io/service: demo
        kuma.io/protocol: tcp
        k8s.io/az: us-east-1
```

Later, based on these tags, the control-plane creates `ClusterLoadAssignment` and puts all tags, for each endpoint, into cluster metadata::

```json
"metadata":{
  "filter_metadata": {
    "envoy.lb": {
      "az/test": "b",
      "kuma.io/protocol": "tcp",
      "k8s.io/az": "us-east-1"
    }
  }
}
```

We can retrieve them from `ClusterLoadAssignment` and match them with the policy. First, we need to retrieve all tags from inbounds so we can determine our location and match it with endpoints.

Pseudo algorithm:

1. Get all inbound tags.
2. Take the matched rules.
3. Check if inbound has the tag defined in the rule. 

> 4. If yes, iterate over all endpoints of the service and group them in to priority groups, by matching with the specific rule(first localZone, then  crossZone). If no, get all the endpoints in the zone.

Reviewer 3: Yes as the last priority. I will clarify this

* [Load Balancer Subsets](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/subsets.html)

1. Priority: Ideal for managing cross-zone traffic and prioritizing the acceptance of specific zones. You can also use it to prioritize local zone endpoints based on these settings.

2. Load Balancer Subsets: Initially considered a good choice for managing localZone traffic. It organizes endpoints into subsets based on provided keys, albeit with a slight increase in Envoy's memory usage. However, I encountered some issues during testing that led me to reconsider this option:

Problems:
* When an endpoint with a matching tag becomes unhealthy, Envoy still attempts to send requests to it and doesn't fall back. This behavior is unacceptable.
* Priorities do not work seamlessly with subsets, even when using aggregated clusters.

3. Locality Weighted Load Balancing: This strategy can help distribute traffic and may be suitable for local zone traffic, but it's not a perfect solution due to the following concerns:

> * We need to deliver weights to the sidecar, which can cause traffic to not stay within the node. Traffic is routed based on these weights, so even if we specify that a node has a higher weight than an availability zone, some traffic might still be routed elsewhere.

Reviewer 3: I feel like it's good but not perfect. It might be difficult to choose the right weight and we are not giving the guarantee that 100% of the traffic stays in the `k8s.io/node` and some requests might get out of the node even if there are enough endpoints to handle it.

We have 2 options:
1. Use only priority
2. Use priority for cross-zone and Locality Weighted Load Balancing for the local zone with different tags. But that won't guarantee that we redirect all the traffic to the same node, but we are going to prefer it.

> * We need to deliver weights to the sidecar, which can cause traffic to not stay within the node. Traffic is routed based on these weights, so even if we specify that a node has a higher weight than an availability zone, some traffic might still be routed elsewhere.

Reviewer 3: I think we should implement second option for now

> * We need to deliver weights to the sidecar, which can cause traffic to not stay within the node. Traffic is routed based on these weights, so even if we specify that a node has a higher weight than an availability zone, some traffic might still be routed elsewhere.

Reviewer: But if you have 10 instances of the service in your zone and only 1 in the same node as you. Do you want to direct 100% of the traffic to this one instance?

I guess it's a tradeoff not clear what's the preferred option

> * We need to deliver weights to the sidecar, which can cause traffic to not stay within the node. Traffic is routed based on these weights, so even if we specify that a node has a higher weight than an availability zone, some traffic might still be routed elsewhere.

Reviewer 3: It's better to not DDos your service, the only thing that I didn't like with the Weighted Load balancing is that when there are not enough endpoints based on the overprovisioning factor, traffic might go to another zone. So even if you have 2 instances out of 3 running in zone an overprovisioning factor might cause some of the traffic can be routed cross zone.

> 

Reviewer 3: https://github.com/kumahq/kuma/pull/8022#discussion_r1365120762

For example:

```yaml
backend_zone1 = SNI backend{mesh=default,kuma.io/zone=zone1}
backend_zone2 = SNI backend{mesh=default,kuma.io/zone=zone2}
```

In addition, it is important to configure the Egress correctly. For each `SNI`, we need to create a filter chain match that sets the required metadata for selecting endpoints:

```yaml
- filterChainMatch:
    serverNames:
    - backend{mesh=default,kuma.io/zone=zone1}
    transportProtocol: tls
  filters:
  - name: envoy.filters.network.tcp_proxy
    typedConfig:
      '@type': type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
      cluster: default:backend
      metadataMatch:
        filterMetadata:
          envoy.lb:
            kuma.io/zone: zone1
      statPrefix: default:backend_zone1
```
Each dataplane has metadata, and it's important to ensure that `kuma.io/zone` is available. If it's not present, add it to the endpoints.

On the ingress side, you only need to adjust the `SNI` to include its corresponding `kuma.io/zone`.

> 

Reviewer: Do we agree that for the first version Egress will manage all clients the same way?

##### Ingress
We don't configure ingresses.

#### ClusterLoadBalancing cache
Currently, while retrieving endpoints in the plugin, we are getting a pointer to the object in the cache that is shared between dataplanes. This won't work with localityAwareness load balancing because each dataplane might have a different subset of endpoints with different priorities.

We are going to implement this without the cache and verify performance.

Steps:
* Run `mesh-perf` test to measure impact of not using CLA cache
* If results are good we'll remove the CLA cache
* Otherwise rewrite CLA cache to keep the different clusters depending on the dataplane tags

##### Test result
We have ran a test of 1000 services with 2 instances each in standalone mode. Each service had reachable services to 3 services defined.

![With CLA Cache](assets/031/xds_generation_with_cache.png)

![Without CLA Cache](assets/031/xds_generation_without_cache.png)

> We can observe that p90 and p99 are much higher without using the cache. We should keep records in the cache but we have to ensure that generated key is reusable so we don't create to many same CLA entries. Also, we could move CLA cache into MeshContext to keep related things together.

Reviewer 3: We could add to the key string of priority to tag and value to make all services that would be in the same location and have the same policy to reuse CLA. Still, there is a chance that we might create much more in case there are many different specific client rules.

`0:k8s.io/node:value,1:k8s.io/az:value,value2,2:kuma.io/zone:zone1`

Reviewer 3: By top targetRef you means 

```
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
```

or 
```
targetRef:
  kind: Mesh
```

Reviewer 3: I don't think it's enough. Because we can have in `to` section `Mesh` and in another case `MeshService` which might provide a different configuration for the service.

Author: we will deal with CLA cache later on as a separate issue

#### MeshHTTPRoute and MeshTCPRoute

> `MeshHTTPRoute` and `MeshTCPRoute` enable the creation of splits for the cluster, and Load Balancing should be applied to them. It's essential to apply Load Balancing before creating the splits. To ensure this order, we must specify that the plugin comes before `Mesh*Route`. Additionally, when generating endpoints for splits, we usually utilize CLA Cache. However, in this scenario, we should retrieve endpoints from the `ResourceSet`.

Reviewer 3: We create a new cluster: `cluster_1_` and add it to the ResourceSet, and then we do the request to the CLA cache with the tags to create CLA for the cluster. Thanks to this we have unique endpoints for this cluster saved in the cache. Next, we add this to ResourceSet. Maybe the order should be after the creation of clusters. I will verify and update docs

# MeshLoadBalancingStrategy - locality aware
* Status: accepted

Technical Story: https://github.com/kumahq/kuma/issues/2689

## Context and Problem Statement
Right now we have really simple locality aware load balancing, we only care about zones. Dataplane will prefer sending traffic to local zone with fallback to other zones. We would like to add more sophisticated mechanism of locality aware load balancing. We would like to allow user to specify load balancing rules based on tags. For example, big cluster is deployed in multiple AZ’s, and we would like to prefer staying in a single AZ to reduce traffic cost.

## Considered Options

## Decision Outcome

> 

Reviewer: Time to update this I think

Because of this we have decided that we will split our API in two sections: `localZone ` and `crossZone`.

Here is the full example of policy:

```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      disabled: false
      localZone:
        affinityTags:                                       # (1)
        - key: "k8s.io/node"                                # (2)
          weight: 90                                        # (3)
        - key: k8s.io/az"
      crossZone:                                            # (4)
        failover:                                           # (5)
          - from:                                           # (6)
              zones: ["zone-1", "zone-2"]
            to:                                             # (7)
              type: Only                                    # (8)
              zones: ["zone-1", "zone-2"]
          - from:
              zones: ["zone-3"]
            to:
              type: AnyExcept
              zones: ["zone-1"]
          - to:
              type: Any
      failoverThreshold: 

>         percentage: 70                                      # (7)

Author: yes, this is configured per cluster

```

(1) In `localZone` section, you configure the affinity of tags used for load balancing. If list is empty or localZone is not specified, load balancing works normally, without taking any priorities into account. 

(2) In `key` field, you define the name of the tag.

(3) In `weight` field, you define proportional load that the group should take. Default is calculated based on the order and might and first in the list is going to have the highest weight while the last the lowest.

Example:
```yaml
affinityTags:
  - key: "k8s.io/node"
  - key: k8s.io/az"
  - key: k8s.io/region"
```

Would create weights:

```yaml

> affinityTags:

Reviewer 4: If I use `weight` does the order become irrelevant? For example, are these 2 equal?
```yaml
affinityTags:
  - key: "k8s.io/node"
    weight: 9
  - key: k8s.io/az"
    weight: 90
  - key: k8s.io/region"
    weight: 900
---
affinityTags:
  - key: k8s.io/region"
  - key: k8s.io/az"
  - key: "k8s.io/node"
```

Author: yes, when weight is configured order becomes irrelevant

  - key: "k8s.io/node"
    weight: 900
  - key: k8s.io/az"
    weight: 90
  - key: k8s.io/region"
    weight: 9
```

and all other endpoints are going to have the weight of 1. When configuring weight you need to either specify weights 

> for all tags or don't specify weights at all.

Author: updated

(4) In `crossZone` section, you configure zones priority.

(5) In `failover` section, you configure list of zone priorities, first rules will have the highest load balancing priority and last in order will have the lowest priority.

> (6) In `from` section, you configure to which zones this rule applies. This rule will apply to all DPPs in `zone-1` and `zone-2`. If not defined is for all zones.

Reviewer 2: do you mean weight on the individual zones in `to` section?

> (6) In `from` section, you configure to which zones this rule applies. This rule will apply to all DPPs in `zone-1` and `zone-2`. If not defined is for all zones.

Reviewer: yes

> (6) In `from` section, you configure to which zones this rule applies. This rule will apply to all DPPs in `zone-1` and `zone-2`. If not defined is for all zones.

Reviewer 3: We don't want to do weights to the zone for now. Zones are based on priority so if you define more zones in the same priority group you can distribute traffic equally. We can think about adding this in the future if someone find it useful

(7) In `to` section, you configure to which zones traffic should be redirected. This field is required.

(8) `type` field lets you control how fallback zones are picked. Allowed values:
- `Only` - will load balance traffic only to zones listed in `zones` field 
- `None` - will not load balance traffic any further
- `Any` - will load balance traffic to all zones
- `AnyExcept` - will load balance traffic to all zones except ones specified in `zones` field

(9) In `failoverThreshold` you configure the percentage of live endpoints below which we will start load balancing to the 
next priority. This is the inverse of [Envoy overprovisioning factor](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/overprovisioning#arch-overview-load-balancing-overprovisioning-factor). 
Default: 50%

### API examples based on use cases

#### Keeping traffic in same zone (disable cross zone traffic)

This:
```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend

>   defaults: {}

Reviewer 4: I'd rather disallow `default: {}` because the result will be whatever is specified in top-level policies after merging. So it's the same as not having the `targetRef` for the `backend` service at all. 

On the other hand, having

```yaml
defaults:
  localityAwareness:
    localZone: 
      affinityTags: []
```

makes much more sense because you're actually overriding the behaviour of the less specific policies

Reviewer 3: Yes, it makes sense. We are going to change it.

- You have have some zone with testing or preview environment and you would like to handle live traffic on it for testing. Is it a use case for locality aware lb? Or should this be done with MeshHTTPRoute?
The question is if we really need it, or can this be done currently in some other way?

### Implementation

We want to enhance current `MeshLoadBalancingStrategy` so code related to the changes should be placed in the plugin.

#### Matching endpoints
Each instance of the service might be in a different locality: zone, node. Users need to properly tag these dataplanes so we are able to match them. We need to retrieve all tags from all inbounds and, later, based on these tags, match outbounds that are in the same location.

Example:

On Kubernetes, all tags from the node could be populated automatically (see details later), while on Universal, users need to populate tags themselves:

```yaml
type: Dataplane
mesh: default
name: demo
networking:
  address: 127.0.0.1
  inbound:
    - port: 9999
      servicePort: 10000
      serviceAddress: 127.0.0.1
      tags:
        kuma.io/service: demo
        kuma.io/protocol: tcp
        k8s.io/az: us-east-1
```

> Later, based on these tags, the control-plane creates `ClusterLoadAssignment` and puts all tags, for each endpoint, into cluster metadata::

Reviewer 2: Yeah, just like we talked in the Slack conversation, putting tags that only matter can actually be a pretty big optimisation imho

> Later, based on these tags, the control-plane creates `ClusterLoadAssignment` and puts all tags, for each endpoint, into cluster metadata::

Author: We will create separate issue for this

```json
"metadata":{
  "filter_metadata": {
    "envoy.lb": {
      "az/test": "b",
      "kuma.io/protocol": "tcp",
      "k8s.io/az": "us-east-1"
    }
  }
}
```

We can retrieve them from `ClusterLoadAssignment` and match them with the policy. First, we need to retrieve all tags from inbounds so we can determine our location and match it with endpoints.

Pseudo algorithm:

1. Get all inbound tags.

> 2. Take the matched rules.

Reviewer 3: I was thinking about matched entries in the `failover` or `affinityTags` because these we need to match. I am going to fix it in the MADR

3. Check if inbound has the tag defined in the rule. 
4. If yes, iterate over all endpoints of the service and group them in to priority groups, by matching with the specific rule(first localZone, then  crossZone). If no, get all the endpoints in the zone. 
   1. Move matching endpoints by first tag to priority 0
   2. Move matching endpoints by n tag to priority  n-1
   3. All not matching endpoint in the local zone move to priority n
   4. Iterate over zone endpoints and based on matching put them in priority N+ iteration
5. Create `LocalityLbEndpoints` with selected endpoints. 

> 6. Override `ClusterLoadAssignment` for the dataplane in the `ResourceSet`.

Reviewer: Currently we do Mesh*Routes and then LaLB:

https://github.com/kumahq/kuma/blob/f3a3e973376967635fbc328948ce7b131ec7f71e/pkg/plugins/policies/policies.go#L21-L23

The routes create the right cluster with the right endpoints in it it seems right?
So sounds like what we need to do is take each cluster and reorganise the endpoints in groups with the right weights right?

This would be cacheable from within the policy I think (the key here would maybe be a little tricky but not impossible).

Reviewer: In any case it really feels like the CLA could be only used in Routes.

Problems:
* When an endpoint with a matching tag becomes unhealthy, Envoy still attempts to send requests to it and doesn't fall back. This behavior is unacceptable.
* Priorities do not work seamlessly with subsets, even when using aggregated clusters.

3. Locality Weighted Load Balancing: This strategy can help distribute traffic and may be suitable for local zone traffic, but it's not a perfect solution due to the following concerns:
* We need to deliver weights to the sidecar, which can cause traffic to not stay within the node. Traffic is routed based on these weights, so even if we specify that a node has a higher weight than an availability zone, some traffic might still be routed elsewhere.

I think we should implement the following:

* Priority: for cross-zone traffic.
* Locality-Weighted Load Balancing: This will help ensure that local zone traffic does not overload instances within the highest priority group when there are more instances available in other groups.

>   * The problem with this approach is that it relies on the health status of hosts, so we need to configure outlier detection and the overprovisioning factor correctly. Ideally, if there are a sufficient number of healthy hosts, the traffic should primarily remain within the local zone and only route outside when there aren't enough healthy hosts. This behavior is largely determined by the overprovisioning factor, which can be set as a default value while allowing users to define it as needed.

Reviewer 2: Why is health a problem? For majority of cases we will just remove the endpoint from EDS before it's unhealthy. For the rest it's actually good that it relies on health, no?

        affinityTags:                                       # (1)
        - key: "k8s.io/node"                                # (2)
          weight: 90                                        # (3)
        - key: k8s.io/az"
          weight: 9
      crossZone:                                            # (4)
        failover:                                           # (5)
          - from:                                           # (6)
              zones: ["zone-1", "zone-2"]
            to:                                             # (7)
              type: Only                                    # (8)
              zones: ["zone-1", "zone-2"]
          - from:
              zones: ["zone-3"]
            to:
              type: AnyExcept
              zones: ["zone-1"]
          - to:
              type: Any
      failoverThreshold: 
        percentage: 70                                      # (7)
```

(1) In `localZone` section, you configure the affinity of tags used for load balancing. If list is empty or localZone is not specified, load balancing works normally, without taking any priorities into account. 

(2) In `key` field, you define the name of the tag.

> (3) In `weight` field, you define proportional load that the group should take. Default is calculated based on the order and might and first in the list is going to have the highest weight while the last the lowest.

Author: fixed

Example:
```yaml
affinityTags:
  - key: "k8s.io/node"
  - key: k8s.io/az"
  - key: k8s.io/region"
```

Would create weights:

```yaml
affinityTags:
  - key: "k8s.io/node"
    weight: 900
  - key: k8s.io/az"
    weight: 90
  - key: k8s.io/region"
    weight: 9

> ```

Author: explained

          - from:
              zones: ["zone-3"]
            to:
              type: AnyExcept
              zones: ["zone-1"]
          - to:
              type: Any
      failoverThreshold: 
        percentage: 70                                      # (7)
```

(1) In `localZone` section, you configure the affinity of tags used for load balancing. If list is empty or localZone is not specified, load balancing works normally, without taking any priorities into account. 

(2) In `key` field, you define the name of the tag.

(3) In `weight` field, you define proportional load that the group should take. Default values are calculated based on the order of the list.
Example:
```yaml
affinityTags:
  - key: "k8s.io/node"
  - key: k8s.io/az"
  - key: k8s.io/region"
```

Would create weights:

```yaml
affinityTags:
  - key: "k8s.io/node"
    weight: 900
  - key: "k8s.io/az"
    weight: 90
  - key: "k8s.io/region"
    weight: 9
```

and all other endpoints are going to have the weight of 1. To compute the percentage of traffic that will stay inside a tag
you need to sum all the weight and then divide weights by it. In this example sum of weights is 999. If you divide weights by it, you 

> get 90% of traffic for `k8s.io/node` 9% for `k8s.io/az` and 1% for `k8s.io/region`. 

Reviewer 3: Defaults are like in the example, so we want to send as much traffic to the first tag but not all. Do you think 90% might be too much?

> get 90% of traffic for `k8s.io/node` 9% for `k8s.io/az` and 1% for `k8s.io/region`. 

Reviewer 3: I've added some more details that is about localzone

- `Only` - will load balance traffic only to zones listed in `zones` field 
- `None` - will not load balance traffic any further
- `Any` - will load balance traffic to all zones
- `AnyExcept` - will load balance traffic to all zones except ones specified in `zones` field

(9) In `failoverThreshold` you configure the percentage of live endpoints below which we will start load balancing to the 
next priority. This is the inverse of [Envoy overprovisioning factor](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/overprovisioning#arch-overview-load-balancing-overprovisioning-factor). 
Default: 50% 
Example: If you have 2 instances in your local zone and suddenly one of instances is unhealthy, traffic is still in the local zone.

### API examples based on use cases

#### Keeping traffic in same zone (disable cross zone traffic)

This:
```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults: 
    localityAwareness: {}
```
and this:
```yaml
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      localZone: 

>         affinityTags: []

Reviewer 5: Does `[]` have different semantics than `null`/leaving out `affinityTags`?

Author: no, all of these cases will work the same

Reviewer 4: the difference will be when merging, right?

Author: yes

On the ingress side, you only need to adjust the `SNI` to include its corresponding `kuma.io/zone`.

However, it's important to note that this step may not be within the scope of the initial implementation. We are going to support initially: **Support only Mesh scope for Egress**.

##### Ingress
We don't configure ingresses.

#### ClusterLoadBalancing cache
Currently, while retrieving endpoints in the plugin, we are getting a pointer to the object in the cache that is shared between dataplanes. This won't work with localityAwareness load balancing because each dataplane might have a different subset of endpoints with different priorities.

We are going to implement this without the cache and verify performance.

Steps:
* Run `mesh-perf` test to measure impact of not using CLA cache
* If results are good we'll remove the CLA cache
* Otherwise rewrite CLA cache to keep the different clusters depending on the dataplane tags

##### Test result
We have ran a test of 1000 services with 2 instances each in standalone mode. Each service had reachable services to 3 services defined.

> ![With CLA Cache](assets/031/xds_generation_with_cache.png)

Reviewer 5: nit: when viewing the MADR, there are no labels indicating which graph is with and which is without

Author: good point, changed

get 90% of traffic for `k8s.io/node` 9% for `k8s.io/az` and 1% for `k8s.io/region`. 
When configuring weight you need to either specify weights for all tags or don't specify weights at all. When you specify weights the order of the list becomes irrelevant.

(4) In `crossZone` section, you configure zones priority.

(5) In `failover` section, you configure list of zone priorities, first rules will have the highest load balancing priority and last in order will have the lowest priority.

(6) In `from` section, you configure to which zones this rule applies. This rule will apply to all DPPs in `zone-1` and `zone-2`. If not defined is for all zones.

(7) In `to` section, you configure to which zones traffic should be redirected. This field is required.

(8) `type` field lets you control how fallback zones are picked. Allowed values:
- `Only` - will load balance traffic only to zones listed in `zones` field 
- `None` - will not load balance traffic any further
- `Any` - will load balance traffic to all zones
- `AnyExcept` - will load balance traffic to all zones except ones specified in `zones` field

> (9) In `failoverThreshold` you configure the percentage of live endpoints below which we will start load balancing to the 

Reviewer 4: Does this field make sense for `localZone`? Should we move this under `crossZone`?

Author: makes sense, changed
